job-manager: support a list of loaded jobtap plugins

grondo opened the pull request:
This WIP PR updates the job-manager jobtap implementation to allow multiple plugins to be a loaded. I'm posting it a bit early in case there is any feedback on the general approach.

Also:

 * A searchpath is now supported (currently set to `$fluxlibdir/job-manager/plugins` for installed flux, of `$abs_top_builddir/src/modules/job-manager/plugins/.libs`). Plugins in the searchpath can be loaded with just their filename (e.g. `flux jobtap load test.so`)
 * Plugins can be specified in a TOML config file via a `[job-manager.plugins]` array (configuration reload is _not_ supported)
 * Since multiple plugins are supported, loaded plugins have to be removed via `flux jobtap remove` or a new `--remove` option to `flux jobtap load`. The remove argument can be a shell glob or `all` to remove all plugins.
 
Still TODO:
 - [x] remove the `builtin.priority.default` plugin since it doesn't do anything
 - [x] change `builtin.priority.hold` to a .so plugin since it won't be commonly used
 - [x] more testing, fix bugs discovered as a result

SteVwonder commented:
After a quick skim of the code, the approach LGTM so far!  One quick question, is there anything in the code currently that prevents loading the same plugin twice?  Nothing stood out to me, but I didn't dive too deeply (or clone and build to try).

grondo commented:
> is there anything in the code currently that prevents loading the same plugin twice? 

No you can currently  load a plugin as many times as you want  :stuck_out_tongue_closed_eyes: 

We'll probably want to fix it so that loading a plugin with the same name unloads the previous plugin with that name, but during development loading the same plugin multiple times was handy for testing.

grondo commented:
I will say that the jobtap interface (for managing plugins) is still a bit rough.

For instance, you can currently easily load a two different plugins that think they are managing job priorities, and I suppose the last one loaded would overwrite priority updates by the first. Similarly, multiple plugins can register handlers for `job.dependency.foo` and both handlers will be called. For that matter if a plugin registers `job.*` it will also get called for `job.dependency.foo` and thus the job-manager wouldn't know that dependency may not have been handle.

My point is that there's still improvements to be done here, and whether we require that now, or push off these enhancements until later is an open question.

grondo commented:
Actually I wonder if would improve usability if the requirement for a separate jobtap plugin name was dropped and plugins were listed by basename? E.g. `dependency-test.so`?

grondo commented:
Ok, I've pushed a checkpoint of work before the weekend. This PR now:

 * changes the default jobtap plugin name to the plugin `basename(3)`. This makes the output of `flux jobtap list` and the argument to `flux jobtap remove` a little more user friendly I think. It is now optional for a jobtap plugin to use `flux_plugin_set_name(3)`. This name is still used for `flux_jobtap_service_register(3)`. If it is not set then it defaults to the soname minus the trailing `.so`.
 * plugins now cannot be loaded multiple times with the same name.

garlick reviewed (commented):
I know you said you were going to re-audit, but was skimming through the code and noticed a couple things and thought I'd mention them in case helpful.

grondo commented on src/modules/job-manager/jobtap.c:
```diff
@@ -740,23 +772,31 @@ void jobtap_handler (flux_t *h,
+        return;
+    if (path && jobtap_handle_load_req (ctx, msg, path, conf) < 0)
+        return;
+    if (flux_respond_pack (h, msg, "{}") < 0)
```
No, there was a payload before, so the call to `flux_respond_pack()` was preserved. I don't think a response payload will be required here so I'll make the suggested change.

grondo commented on src/modules/job-manager/jobtap.c:
```diff
@@ -870,35 +871,79 @@ static zlistx_t *path_list (const char *searchpath, const char *path)
+    if ((q = strchr (copy, '.')))
+        *q = '\0';
+    rc = flux_plugin_set_name (p, copy);
+    free (copy);
```
I tried to avoid returning errno-only reponses and include a useful error message in all responses instead. Not sure an errno from plugin_set_name() will be useful to the caller, but I'll try to preserve it anyway.

garlick replied:
Yeah, I guess the value of errno doesn't matter that much if there's a human readable error.  We just need to ensure errno != 0 or `flux_respond_error()` will fail and not send the response at all.  (In case that wasn't clear)

grondo replied:
Ah, good point! I didn't know that or had forgotten it. I wonder if that is overall desired behavior now that `flux_respond_error()` can take an error message?

grondo replied:
Regardless, I'll try to ensure errno is preserved here

garlick replied:
I struggled with that one in that now-defunct EFLUXED PR, because in some cases you don't know if the caller will do the extra work to decode a human readable error, so returning errnum=0 or a misleading value is not great and should be detected by unit tests.

In this case the RPC is not really intended to be used by anything other than the plugin loader, so you could just call `flux_respond_error (h, msg, EINVAL, errtxt)` and just make sure errtxt is non-NULL?

grondo replied:
Ah, good point. I'll try to preserve errno but also do something like this?
```C
  flux_respond_error (h, msg, errno != 0 ? errno : EINVAL, errmsg)
```

garlick replied:
Yeah that might be fine in this case.

grondo commented:
> thought I'd mention them in case helpful.

Thanks, that is helpful. I find it difficult to get started auditing my own code that has gone through multiple iterations!

grondo commented:
Ok, I've made the changes suggested by @garlick and squashed the result. I've removed WIP as this is ready for a more thorough review.

garlick commented on src/modules/job-manager/plugins/priority-hold.c:
```diff
@@ -31,7 +31,9 @@ static int hold_cb (flux_plugin_t *p,
 }
```
Comment needs an update (no builtin plugins, and basename change)
```c
/* builtins/hold.c - builtin plugin that holds all submitted job
```

garlick reviewed (commented):
Couple more random comments before a closer review:

I made a little hello world jobtap plugin that subscribes to "*" and prints stuff.  Whee! :-)

This can be removed from `flux-jobtap-plugins(7)`:
> NOTE:  Currently only a single jobtap plugin may be loaded into the job
              manager at a time.

Also the mention of "builtin and external plugins", since builtins are gone.

Should anything be added to the man page to describe how e.g. multiple loaded jobtap plugins interact?

Should there be a reprioritization pass when a plugin is unloaded to re-establish the default or formerly shadowed plugin priorities?  Case in point: load `priority-hold.so`, submit job (now held), unload `priority-hold.so` - should the job be "un-held"?

grondo commented:
> Case in point: load priority-hold.so, submit job (now held), unload priority-hold.so - should the job be "un-held"?

Good question. I thought about this and I didn't think so. Whe reloading a plugin it is first unloaded, then loaded. If reloading the priority-hold.so plugin, I don't think a user intends to release all jobs. Even if simply unloading the priority-hold.so plugin, the least surprising result is probably that all previously held jobs stay that way. Loading a _new_ plugin that sets a priority, however, would release jobs (and there is a test to this effect.)

(Does that make any sense?)

> Also the mention of "builtin and external plugins", since builtins are gone.

Thanks for the reminder about the manpage! 

builtins still exist, there just aren't any yet, but a future PR will probably add `begin-time` and `after` builtin dependency plugins. I'll make a pass at the rest of the manpage.

garlick commented:
Ah, gotcha.  That all makes sense!

grondo commented:
Oh, I was also wondering if the commit to ignore DSOs without a `mod_name` symbol in `flux_modfind(3)` should go in a separate PR for release notes?

garlick commented:
> if the commit to ignore DSOs without a mod_name symbol in flux_modfind(3) should go in a separate PR for release notes?

That might be a good idea, although I don't have a strong opinion if you'd rather avoid the extra work.

garlick commented on src/modules/job-manager/jobtap.c:
```diff
@@ -67,6 +67,26 @@ struct dependency {
                                           struct job *job);
 
+
+static int errprintf (jobtap_error_t *errp, const char *fmt, ...)
```
Nice cleanup.  It might be a good idea to put one of these in libutil along with a generic `struct error` for internal reuse, but it's fine if we save that for later.  Maybe we can can come up with something that helps us preserve errno too.

Anyway, this looks good.  Don't change.

garlick commented on src/modules/job-manager/jobtap.c:
```diff
@@ -219,6 +183,216 @@ static flux_plugin_arg_t *jobtap_args_vcreate (struct jobtap *jobtap,
+
+    /*  Make plugin aware of all active jobs via job.new callback
+     */
+    jobs = zhashx_values (ctx->active_jobs);
```
check for NULL return

garlick commented on src/modules/job-manager/jobtap.c:
```diff
@@ -219,6 +183,216 @@ static flux_plugin_arg_t *jobtap_args_vcreate (struct jobtap *jobtap,
+                          "{s?:{s?:o}}",
+                          "job-manager",
+                            "plugins", &plugins) < 0) {
+        return errprintf (errp,
+                          "[job-manager.plugins]: %s: %d: %s",
+                          error.filename,
+                          error.lineno,
+                          error.errbuf);
```
Not sure filename and lineno are going to be helpful here after file glob -> TOML -> JSON conversion,.

grondo replied:
Good point.

garlick reviewed (approved):
Made a review pass and just noted a few minor things.  Generally looks great!

After updating man page I'm fine if this goes in so marking approved.

grondo commented:
Ok, I fixed and squashed the issues commented by @garlick above. I also added a `flux-jobtap(1)` manpage so it can be referenced by the `flux-jobtap-plugins(7)` manpage when describing a method to load, query and remove plugins.

A section on the new `[job-manager.plugins]` configuration array was also added to `flux-jobtap-plugins(7)`.

grondo commented:
We should probably add manpages for the `flux_jobtap_*(3)` calls, but wasn't sure that was necessary here quite yet. I'll be happy to work one up if required, but for now I'll open an issue to get that done at some point (maybe once we're sure we have the right interfaces there)

garlick commented:
> We should probably add manpages for the flux_jobtap_*(3) calls, but wasn't sure that was necessary here quite yet. I'll be happy to work one up if required, but for now I'll open an issue to get that done at some point (maybe once we're sure we have the right interfaces there)

Makes sense to me to wait till things have settled.

garlick commented:
Man pages look good!  I'm fine with this going in whenever you are ready.

grondo commented:
Ok, I've been building more features on top of this branch without any issues so setting MWP.
